tools/manifest/manifest.py

```diff
@@ -285,119 +155,107 @@ def update(self, tree):
         unusual API is designed as an optimistaion meaning that SourceFile items need not be
         constructed in the case we are not updating a path, but the absence of an item from
         the iterator may be used to remove defunct entries from the manifest."""
-        seen_files = set()  # type: Set[Text]
 
         changed = False
 
         # Create local variable references to these dicts so we avoid the
         # attribute access in the hot loop below
-        path_hash = self._path_hash  # type: Dict[Text, Tuple[Text, Text]]
         data = self._data
 
-        prev_files = data.paths()  # type: Set[Text]
+        types = data.types()
+        deleted = set(types)
```

Comment:
So creating a set from a the `dict`'s keys is much quicker than creating it from our underlying data structure, because it then does the entire loop in C, rather than having to call our Python `__iter__` implementation.

Given we need both `types` and `deleted` here, we really do only want to iterate over `TypeData` once, because otherwise that adds significant cost. I guess I should add a comment to that effect?

(And renamed it to `type_by_path`)

Arguably the other question here is whether we even need `type_by_path`; we only look up `not is_new` tests, and mostly we only need it to find the old hash. It avoids us having to iterate through all of the test types, but it's not clear that we actually save enough but computing it ahead-of-time for the average number of tests modified.

Reply:
I'm happy with this for now; I think further changes should come with profiling.